### PR TITLE
Convert image and fefault to webp

### PR DESCRIPTION
**Pull request by guyyoo**
https://github.com/digi-serve/ab_service_file_processor/issues/47
## Release Notes

- Convert image and fefault to webp.

### CODE REVIEW

File: utils/pathUtils.js [hunk 1]
```diff
@@ -7,6 +7,20 @@ const fs = require("fs");
...
 
 module.exports = {
+   /**
+    * @function destPath
+    * return true if the path exists.
+    * @param {string} destPath
+    * @return {Promise<boolean>}
+    */
+   checkPath: async (destPath) =>
+      await new Promise((resolve) => {
```
Comment (wongpratan) [inline utils/pathUtils.js:17]: Do we need `return` here ?

File: utils/imageUtils.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+         case "heic":
+         case "jpg":
+         case "png":
+         case "tiff":
+         case "webp":
+            break;
+         default:
+            return;
+      }
+      const args = ["convert"];
```
Comment (wongpratan) [inline utils/imageUtils.js:23]: Is this above code checking extension type ?

File: utils/pathUtils.js (hunk 1 above)
Comment (guyyoo) [reply utils/pathUtils.js:17]: No, because I don't put `{}` after `=>`.

File: utils/imageUtils.js (hunk 2 above)
Comment (guyyoo) [reply utils/imageUtils.js:23]: Yes, It is.

**Review by wongpratan — APPROVED** [review]

File: utils/imageUtils.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * files.
+ */
+const { exec } = require("child_process");
+const fs = require("fs");
+const { reject } = require("lodash");
+const path = require("path");
+
+module.exports = {
+   convert: async (imagePath, extension, quality, resize) => {
+      switch (extension) {
```
Comment (nh758) [inline utils/imageUtils.js:13]: suggest:
```
const validExtensions = ["heic", "jpg", ...];
if (!validExtendions.includes(extension)) return;
```

File: handlers/file-get.js [hunk 4]
```diff
@@ -2,11 +2,49 @@
...
+const pathUtils = require("../utils/pathUtils.js");
+
+const getFilePath = async (req, file, needOriginalImage) => {
+   let filePath = file.pathFile;
+   switch (file.type) {
+      case "image/heic":
+      case "image/jpeg":
+      case "image/png":
+      case "image/tiff":
+      case "image/webp":
```
Comment (nh758) [inline handlers/file-get.js:19]: i think we can just return the main filepath if it's already a `webp`?

File: handlers/file-get.js [hunk 5]
```diff
@@ -2,11 +2,49 @@
...
+         const imageWEBPPath = path.join(
+            parsedImagePath.dir,
+            `${parsedImagePath.name}.webp`
+         );
+         if ((await pathUtils.checkPath(imageWEBPPath))) {
+            if (!needOriginalImage) filePath = imageWEBPPath;
+            break;
+         }
+         if (await pathUtils.checkPath(filePath))
+            req.worker(
```
Comment (nh758) [inline handlers/file-get.js:30]: I think we should also convert after a file upload.

**Review by nh758 — CHANGES_REQUESTED** [review]
I think we should also trigger the conversion after the file is uploaded. We should still have it in the file-get for files already on the server.

File: utils/imageUtils.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      const args = ["convert"];
+      if (quality != null) args.push(`-quality ${quality}`);
+      if (resize != null) args.push(`-resize ${resize}`);
+      args.push(imagePath);
+      const parsedImagePath = path.parse(imagePath);
+      args.push(path.join(parsedImagePath.dir, `${parsedImagePath.name}.${extension}`));
+      await new Promise((resolve) => {
+         exec(args.join(" "), (error) => {
+            if (error) reject(error);
```
Comment (Hiro-Nakamura) [inline utils/imageUtils.js:31]: what `reject()` does this refer to?

File: utils/imageUtils.js (hunk 3 above)
Comment (Hiro-Nakamura) [reply utils/imageUtils.js:13]: I like this.  
Also, if we could keep this someplace more central like: `AB.objectFile().validExtensions`  so we can keep 1 copy of this info and use it multiple places.

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
OK, I like what you have done.  It will start working immediately and we don't have to worry about doing any conversion on existing images that are uploaded.

But I have a feeling, 3-6 months from now, if you asked me to take at look at the code that converts to `webp`, I wouldn't think to look in the `file-get` handler.  To me it makes more sense long term to be performing the conversions when we upload the file.  It might be worth a discussion to see if it makes better sense long term, to have this functionality here as opposed to the `image-upload` handler.

Also, we are using `req.worker()`  to offload the image conversion logic.  This would make perfect sense if we were using a javascript package that performs the conversion (like `jimp`).  But in this case, we are using `exec()` to call the command line tool `imagemagick` to do the processing.  This already creates another thread for us.  So doing `exec()` within the `req.worker()` doesn't gain us anything and just makes things a little more complicated.

File: utils/imageUtils.js (hunk 3 above)
Comment (guyyoo) [reply utils/imageUtils.js]: https://github.com/digi-serve/appbuilder_platform_service/pull/90

File: handlers/file-get.js (hunk 5 above)
Comment (guyyoo) [reply handlers/file-get.js]: https://github.com/digi-serve/ab_service_file_processor/pull/52/commits/a68698dfafb5fd42c982db2535efaf36eef2c54a
https://github.com/digi-serve/ab_service_api_sails/pull/120

[the remark this answers is not in the document]
Comment (guyyoo) [reply handlers/file-get.js]: https://github.com/digi-serve/ab_service_file_processor/pull/52/commits/a68698dfafb5fd42c982db2535efaf36eef2c54a
https://github.com/digi-serve/ab_service_api_sails/pull/120

File: handlers/file-get.js (hunk 4 above)
Comment (guyyoo) [reply handlers/file-get.js]: https://github.com/digi-serve/ab_service_file_processor/pull/52/commits/a68698dfafb5fd42c982db2535efaf36eef2c54a
https://github.com/digi-serve/ab_service_api_sails/pull/120

**Review by nh758 — APPROVED** [review]

File: handlers/file-get.js [hunk 7]
```diff
@@ -47,36 +51,65 @@ module.exports = {
...
+            case "image/tiff":
+            case "image/webp":
+               {
+                  const newFileExtension = req.param("extension") || "webp";
+                  const newFilePath = path.join(
+                     parsedFilePath.dir,
+                     `${parsedFilePath.name}.${newFileExtension}`
+                  );
+                  if (!fileObject.validExtension(newFilePath))
+                     throw new Error(`The file extension "${newFileExtension}" is invalid.`);
```
Comment (Hiro-Nakamura) [inline handlers/file-get.js:93]: should update `errorContext` before throwing an error.

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
Simple update then Good To Go

**Pull request merged**